store/plugins/userPlugin.ts

```diff
@@ -0,0 +1,106 @@
+import { Plugin } from 'vuex';
+import { DisplayPreferencesDto } from '@jellyfin/client-axios';
+import destr from 'destr';
+import { AppState } from '..';
+import {
+  ClientSettingsState,
+  getDefaultState as settingState
+} from '~/store/clientSettings';
+
+export interface ClientPreferences
+  extends Omit<DisplayPreferencesDto, 'CustomPrefs'> {
+  CustomPrefs?: ClientSettingsState;
+}
+
+export interface DisplayPreferencesApiState {
+  syncing: boolean;
+  LastSync: number;
+  LastSettingChange: number;
+}
+
+/**
+ * Cast custom preferences returned from the server from strings to the correct Javascript type
+ *
+ * @param {DisplayPreferencesDto} data - Response from the server
+ * @returns {ClientPreferences} - The response with the correct datatypes.
+ */
+function castResponse(data: DisplayPreferencesDto): ClientPreferences {
+  const result = data as ClientPreferences;
+  for (const [key, value] of Object.entries(
+    result.CustomPrefs as ClientSettingsState
+  )) {
+    /**
+     * destr does proper conversion for all the types, even undefined
+     */
+    // @ts-expect-error - TypeScript can't infer types from Object.entries
+    result.CustomPrefs[key] = destr(value);
+  }
+  return result;
+}
```

Comment:
I'm not sure all of the functions for casting should live and parsing should live here. Imo, they should be where all the "magic" happens, and that's the ``preferencesSyncPlugin``

I don't know if that's supported server-side right now but, ideally, if I change one setting in another session, the websocket would inform me of that change, so I could pull those changes and get them. This means that the fetching part will not only happen in login, but also while the sesion is active.

That's why I would remove this plugin altogether and move all the logic to the ``preferencesSync`` one. It makes sense all the syncing is there, this being ``userPlugin`` is just confusing imo (in fact took me a while to understand how everything worked).

For now, no need to worry about the websocket-informing-of-changes stuff, but we could move all the logic there so the future refactor is minimal.

Reply:
> I'm not sure all of the functions for casting should live and parsing should live here. Imo, they should be where all the "magic" happens, and that's the preferencesSyncPlugin

That plugin doesn't do any casting, since it only pushed data. This should live where it's used, so in this file.

> That's why I would remove this plugin altogether and move all the logic to the `preferencesSync` one. It makes sense all the syncing is there, this being `userPlugin` is just confusing imo (in fact took me a while to understand how everything worked).

This is a generic user authentication plugin. Anything that needs something done on login/logout would be done here.

It just contains preference fetching and the store reset atm, because that's all we're doing at the moment in these cases.

The Vuex plugins should be organized by context, not by intent per-se. So anything that deals with the websocket goes in that pllugin, anything that goes with auth goes in the user plugin, etc.